### PR TITLE
feat($moduleName): extract the class module names

### PR DESCRIPTION
**Pull request by dweinstein**
~~- WIP / untested~~
- [x] test broken
- [x] test

### CODE REVIEW

**Comment by dweinstein** [comment]
Currently having trouble testing as I get the following perplexing error on my box:

```
± make -C test
... hamsters ...
build/macos-x86_64/runner
This copy of libswiftCore.dylib requires an OS version prior to 10.14.4.
make: *** [run-macos] Abort trap: 6

```

File: index.js [hunk 1]
```diff
@@ -580,6 +582,12 @@ function Runtime() {
                                 cachedIvars = new ObjCIvars(self, classHandle());
                         }
                         return cachedIvars;
+                    case "$imageName":
+                        if (cachedImageName === null) {
+                            if (isClass())
+                                cachedImageName = api.class_getImageName(handle);
```
Comment (oleavr) [inline index.js]: Perhaps we could remove the `if` and use `classHandle()` instead of `handle`?

**Comment by oleavr** [comment]
Should consider naming it `$moduleName` to use Frida-terminology.

Once this lands we should update the:
- [ ] [Reference docs](https://github.com/frida/frida-website/blob/master/_docs/javascript-api.md)
- [ ] [TypeScript bindings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/598ffd0b99dcfa511ab6056d9652c99665a7bbe4/types/frida-gum/index.d.ts#L3345)

File: index.js (hunk 1 above)
Comment (dweinstein) [reply index.js]: will do, also found bug while writing the test :-)

**Comment by dweinstein** [comment]
> Should consider naming it `$moduleName` to use Frida-terminology.
Done in https://github.com/frida/frida-objc/commit/cede9f6a73552f7c9186fd6c3180c8068c4d93c1

> Once this lands we should update the:
> 
> * [X]  [Reference docs](https://github.com/frida/frida-website/blob/master/_docs/javascript-api.md) (done) https://github.com/frida/frida-website/pull/137

> * [ ]  [TypeScript bindings](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/598ffd0b99dcfa511ab6056d9652c99665a7bbe4/types/frida-gum/index.d.ts#L3345)

Not sure best way to handle this one

File: test/basics.m [hunk 2]
```diff
@@ -296,6 +297,15 @@ - (void)secret {}
   EXPECT_NO_MESSAGES ();
 }
 
+TESTCASE (module_name_can_be_retrieved)
+{
+  COMPILE_AND_LOAD_SCRIPT (
+      "var NSString = ObjC.classes.NSString;"
+      "var badger = NSString.stringWithString_(\"badger\");"
+      "send(badger.$moduleName);");
+  EXPECT_SEND_MESSAGE_WITH ("\"/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation\"");
```
Comment (oleavr) [inline test/basics.m]:
```suggestion
  EXPECT_SEND_MESSAGE_WITH ("\"/System/Library/Frameworks/"
      "CoreFoundation.framework/Versions/A/CoreFoundation\"");
```

Comment (dweinstein) [reply test/basics.m]: Interestingly the path changes between `NSString.$moduleName` and the `badger` -- updated test incoming w/ de39cae

**Comment by oleavr** [comment]
Nit: perhaps we should move it right after `$className` like you did in the docs?

File: index.js [hunk 3]
```diff
@@ -471,6 +473,11 @@ function Runtime() {
                         return function () {
                             return receiver.$className;
                         };
+                    case "$moduleName":
+                        if (cachedModuleName === null) {
+                            cachedModuleName = api.class_getImageName(classHandle()).readUtf8String();
+                        }
+                        return cachedModuleName;
```
Comment (oleavr) [inline index.js]: I guess it should move down to right after `case "$className"`?

File: index.js [hunk 4]
```diff
@@ -429,6 +430,7 @@ function Runtime() {
         let cachedOwnMethodNames = null;
         let cachedIvars = null;
         let weakRef = null;
+        let cachedModuleName = null;
```
Comment (oleavr) [inline index.js]: Should follow `cachedClassName` I'd say.

File: index.js [hunk 5]
```diff
@@ -2677,6 +2684,7 @@ function getApi() {
                 "class_copyProtocolList": ['pointer', ['pointer', 'pointer']],
                 "class_copyMethodList": ['pointer', ['pointer', 'pointer']],
                 "class_getClassMethod": ['pointer', ['pointer', 'pointer']],
+                "class_getImageName": ['pointer', ['pointer']],
```
Comment (oleavr) [inline index.js]: Should go right after `class_getName`. (Ordering isn't currently alphabetical.)

**Pull request merged**